/* OpenAPIArt Test API 0.0.1
 * Demonstrates the features of the OpenAPIArt package
 * License: NO-LICENSE-PRESENT */

syntax = "proto3";

package sanity;

option go_package = "./sanity;sanity";

import "google/protobuf/descriptor.proto";
import "google/protobuf/empty.proto";

// Description missing in models
message ErrorDetails {

  // Description missing in models
  repeated string errors = 1;
}

// Description missing in models
message WarningDetails {

  // Description missing in models
  repeated string warnings = 1;
}

// Error response generated while serving API request.
message Error {

  // Numeric status code based on underlying transport being used.
  // required = true
  int32 code = 1;

  message Kind {
    enum Enum {
      unspecified = 0;
      transport = 1;
      validation = 2;
      internal = 3;
    }
  }
  // Kind of error message.
  optional Kind.Enum kind = 2;

  // List of error messages generated while serving API request.
  repeated string errors = 3;
}

// Container which retains the configuration
message PrefixConfig {

  // A required object that MUST be generated as such.
  // required = true
  EObject required_object = 1;

  // An optional object that MUST be generated as such.
  optional EObject optional_object = 2;

  // Description missing in models
  optional bool ieee_802_1qbb = 3;

  // Deprecated: Information TBD
  // 
  // Description TBD
  optional int32 space_1 = 4;

  // Description missing in models
  optional int64 full_duplex_100_mb = 5;

  message Response {
    enum Enum {
      unspecified = 0;
      status_200 = 1;
      status_400 = 2;
      status_404 = 3;
      status_500 = 4;
    }
  }
  // Indicate to the server what response should be returned
  // default = Response.Enum.status_200
  optional Response.Enum response = 6;

  // Under Review: Information TBD
  // 
  // Small single line description
  // required = true
  string a = 7;

  // Longer multi-line description
  // Second line is here
  // Third line
  // required = true
  float b = 8;

  // Description missing in models
  // required = true
  int32 c = 9;

  message DValues {
    enum Enum {
      unspecified = 0;
      a = 1;
      b = 2;
      c = 3;
    }
  }
  // Deprecated: Information TBD
  // 
  // A list of enum values
  repeated DValues.Enum d_values = 10;

  // Deprecated: Information TBD
  // 
  // A child object
  optional EObject e = 11;

  // An object with only choice(s)
  optional FObject f = 12;

  // A list of objects with choice and properties
  repeated GObject g = 13;

  // A boolean value
  // default = True
  optional bool h = 14;

  // A byte string
  optional bytes i = 15;

  // A list of objects with only choice
  repeated JObject j = 16;

  // A nested object with only one property which is a choice object
  optional KObject k = 17;

  // Description missing in models
  optional LObject l = 18;

  // A list of string values
  repeated string list_of_string_values = 19;

  // A list of integer values
  repeated int32 list_of_integer_values = 20;

  // Description missing in models
  optional LevelOne level = 21;

  // Description missing in models
  optional Mandate mandatory = 22;

  // Description missing in models
  optional Ipv4Pattern ipv4_pattern = 23;

  // Description missing in models
  optional Ipv6Pattern ipv6_pattern = 24;

  // Description missing in models
  optional MacPattern mac_pattern = 25;

  // Description missing in models
  optional IntegerPattern integer_pattern = 26;

  // Description missing in models
  optional ChecksumPattern checksum_pattern = 27;

  // Description missing in models
  optional Layer1Ieee802x case = 28;

  // Description missing in models
  optional MObject m_object = 29;

  // int64 type
  optional int64 integer64 = 30;

  // int64 type list
  repeated int64 integer64_list = 31;

  // Description missing in models
  optional PatternPrefixConfigHeaderChecksum header_checksum = 32;

  // Under Review: Information TBD
  // 
  // string minimum&maximum Length
  optional string str_len = 33;

  // Under Review: Information TBD
  // 
  // Array of Hex
  repeated string hex_slice = 34;

  // Description missing in models
  optional PatternPrefixConfigAutoFieldTest auto_field_test = 35;

  // Description missing in models
  optional string name = 36;

  // Description missing in models
  repeated WObject w_list = 37;

  // Description missing in models
  repeated ZObject x_list = 38;

  // Description missing in models
  optional ZObject z_object = 39;

  // Description missing in models
  optional YObject y_object = 40;

  // A list of objects with choice with and without properties
  repeated ChoiceObject choice_object = 41;

  // Description missing in models
  optional RequiredChoiceParent required_choice_object = 42;

  // A list of objects with choice and properties
  repeated GObject g1 = 43;

  // A list of objects with choice and properties
  repeated GObject g2 = 44;

  // int32 type
  optional int32 int32_param = 45;

  // int32 type list
  repeated int32 int32_list_param = 46;

  // uint32 type
  optional uint32 uint32_param = 47;

  // uint32 type list
  repeated uint32 uint32_list_param = 48;

  // uint64 type
  optional uint64 uint64_param = 49;

  // uint64 type list
  repeated uint64 uint64_list_param = 50;

  // should automatically set type to int32
  optional int32 auto_int32_param = 51;

  // should automatically set type to []int32
  repeated int32 auto_int32_list_param = 52;
}

// Description missing in models
message WObject {

  // Description missing in models
  // required = true
  string w_name = 1;
}

// Description missing in models
message ZObject {

  // Description missing in models
  // required = true
  string name = 1;
}

// Description missing in models
message YObject {

  // TBD
  // 
  // x-constraint:
  // - /components/schemas/ZObject/properties/name
  // - /components/schemas/WObject/properties/w_name
  // 
  optional string y_name = 1;
}

// Description missing in models
message Layer1Ieee802x {

  // Description missing in models
  optional bool flow_control = 1;
}

// Deprecated: new schema Jobject to be used
// 
// Description TBD
message GObject {

  // Description missing in models
  // default = asdf
  optional string g_a = 1;

  // Description missing in models
  // default = 6
  optional int32 g_b = 2;

  // Deprecated: Information TBD
  // 
  // Description TBD
  // default = 77.7
  optional float g_c = 3;

  message Choice {
    enum Enum {
      unspecified = 0;
      g_d = 1;
      g_e = 2;
    }
  }
  // Description missing in models
  // default = Choice.Enum.g_d
  optional Choice.Enum choice = 4;

  // Description missing in models
  // default = some string
  optional string g_d = 5;

  // Description missing in models
  // default = 3.0
  optional double g_e = 6;

  message GF {
    enum Enum {
      unspecified = 0;
      a = 1;
      b = 2;
      c = 3;
    }
  }
  // Another enum to test protbuf enum generation
  // default = GF.Enum.a
  optional GF.Enum g_f = 7;

  // Description missing in models
  optional string name = 8;
}

// Description missing in models
message EObject {

  // Description missing in models
  // required = true
  float e_a = 1;

  // Description missing in models
  // required = true
  double e_b = 2;

  // Description missing in models
  optional string name = 3;

  // Description missing in models
  optional string m_param1 = 4;

  // Description missing in models
  optional string m_param2 = 5;
}

// Description missing in models
message FObject {

  message Choice {
    enum Enum {
      unspecified = 0;
      f_a = 1;
      f_b = 2;
      f_c = 3;
    }
  }
  // Description missing in models
  // default = Choice.Enum.f_a
  optional Choice.Enum choice = 1;

  // Description missing in models
  // default = some string
  optional string f_a = 2;

  // Description missing in models
  // default = 3.0
  optional double f_b = 3;
}

// Description missing in models
message JObject {

  message Choice {
    enum Enum {
      unspecified = 0;
      j_a = 1;
      j_b = 2;
    }
  }
  // Description missing in models
  // default = Choice.Enum.j_a
  optional Choice.Enum choice = 1;

  // Description missing in models
  optional EObject j_a = 2;

  // Description missing in models
  optional FObject j_b = 3;
}

// Description missing in models
message ChoiceObject {

  message Choice {
    enum Enum {
      unspecified = 0;
      e_obj = 1;
      f_obj = 2;
      no_obj = 3;
    }
  }
  // Description missing in models
  // default = Choice.Enum.no_obj
  optional Choice.Enum choice = 1;

  // Description missing in models
  optional EObject e_obj = 2;

  // Description missing in models
  optional FObject f_obj = 3;
}

// Description missing in models
message KObject {

  // Description missing in models
  optional EObject e_object = 1;

  // Description missing in models
  optional FObject f_object = 2;
}

// Format validation object
message LObject {

  // Description missing in models
  optional string string_param = 1;

  // Description missing in models
  optional int32 integer = 2;

  // Description missing in models
  optional float float = 3;

  // Description missing in models
  optional double double = 4;

  // Description missing in models
  optional string mac = 5;

  // Description missing in models
  optional string ipv4 = 6;

  // Description missing in models
  optional string ipv6 = 7;

  // Description missing in models
  optional string hex = 8;
}

// Required format validation object
message MObject {

  // Description missing in models
  // required = true
  string string_param = 1;

  // Description missing in models
  // required = true
  int32 integer = 2;

  // Description missing in models
  // required = true
  float float = 3;

  // Description missing in models
  // required = true
  double double = 4;

  // Description missing in models
  // required = true
  string mac = 5;

  // Description missing in models
  // required = true
  string ipv4 = 6;

  // Description missing in models
  // required = true
  string ipv6 = 7;

  // Description missing in models
  // required = true
  string hex = 8;
}

// Object to Test required Parameter
message Mandate {

  // Description missing in models
  // required = true
  string required_param = 1;
}

// Under Review: the whole schema is being reviewed
// 
// Object to Test required Parameter
message UpdateConfig {

  // A list of objects with choice and properties
  repeated GObject g = 1;
}

// Description missing in models
message MetricsRequest {

  message Choice {
    enum Enum {
      unspecified = 0;
      port = 1;
      flow = 2;
    }
  }
  // Description missing in models
  // default = Choice.Enum.port
  optional Choice.Enum choice = 1;

  // Description missing in models
  optional string port = 2;

  // Description missing in models
  optional string flow = 3;
}

// Description missing in models
message Metrics {

  message Choice {
    enum Enum {
      unspecified = 0;
      ports = 1;
      flows = 2;
    }
  }
  // Description missing in models
  // default = Choice.Enum.ports
  optional Choice.Enum choice = 1;

  // Description missing in models
  repeated PortMetric ports = 2;

  // Description missing in models
  repeated FlowMetric flows = 3;
}

// Description missing in models
message PortMetric {

  // Description missing in models
  // required = true
  string name = 1;

  // Description missing in models
  // required = true
  double tx_frames = 2;

  // Description missing in models
  // required = true
  double rx_frames = 3;
}

// Description missing in models
message FlowMetric {

  // Description missing in models
  // required = true
  string name = 1;

  // Description missing in models
  // required = true
  double tx_frames = 2;

  // Description missing in models
  // required = true
  double rx_frames = 3;
}

// Description missing in models
message RequiredChoiceParent {

  message Choice {
    enum Enum {
      unspecified = 0;
      intermediate_obj = 1;
      no_obj = 2;
    }
  }
  // Description missing in models
  // required = true
  Choice.Enum choice = 1;

  // Description missing in models
  optional RequiredChoiceIntermediate intermediate_obj = 2;
}

// Description missing in models
message RequiredChoiceIntermediate {

  message Choice {
    enum Enum {
      unspecified = 0;
      f_a = 1;
      leaf = 2;
    }
  }
  // Description missing in models
  // required = true
  Choice.Enum choice = 1;

  // Description missing in models
  // default = some string
  optional string f_a = 2;

  // Description missing in models
  optional RequiredChoiceIntermeLeaf leaf = 3;
}

// Description missing in models
message RequiredChoiceIntermeLeaf {

  // Description missing in models
  optional string name = 1;
}

// To Test Multi level non-primitive types
message LevelOne {

  // Level one
  optional LevelTwo l1_p1 = 1;

  // Level one to four
  optional LevelFour l1_p2 = 2;
}

// Test Level 2
message LevelTwo {

  // Level Two
  optional LevelThree l2_p1 = 1;
}

// Test Level3
message LevelThree {

  // Set value at Level 3
  optional string l3_p1 = 1;
}

// Test level4 redundant junk testing
message LevelFour {

  // loop over level 1
  optional LevelOne l4_p1 = 1;
}

// Test ipv4 pattern
message Ipv4Pattern {

  // Description missing in models
  optional PatternIpv4PatternIpv4 ipv4 = 1;
}

// Test ipv6 pattern
message Ipv6Pattern {

  // Description missing in models
  optional PatternIpv6PatternIpv6 ipv6 = 1;
}

// Test mac pattern
message MacPattern {

  // Description missing in models
  optional PatternMacPatternMac mac = 1;
}

// Test integer pattern
message IntegerPattern {

  // Description missing in models
  optional PatternIntegerPatternInteger integer = 1;
}

// Test checksum pattern
message ChecksumPattern {

  // Description missing in models
  optional PatternChecksumPatternChecksum checksum = 1;
}

// Description missing in models
message CommonResponseSuccess {

  // Description missing in models
  optional string message = 1;
}

// Description missing in models
message ApiTestInputBody {

  // Description missing in models
  optional string some_string = 1;
}

// Description missing in models
message ServiceAbcItem {

  // Description missing in models
  optional string some_id = 1;

  // Description missing in models
  optional string some_string = 2;

  // Description missing in models
  optional string path_id = 3;

  // Description missing in models
  optional string level_2 = 4;
}

// Description missing in models
message ServiceAbcItemList {

  // Description missing in models
  repeated ServiceAbcItem items = 1;
}

// Header checksum
message PatternPrefixConfigHeaderChecksum {

  message Choice {
    enum Enum {
      unspecified = 0;
      generated = 1;
      custom = 2;
    }
  }
  // The type of checksum
  // default = Choice.Enum.generated
  optional Choice.Enum choice = 1;

  message Generated {
    enum Enum {
      unspecified = 0;
      good = 1;
      bad = 2;
    }
  }
  // A system generated checksum value
  // default = Generated.Enum.good
  optional Generated.Enum generated = 2;

  // A custom checksum value
  optional uint32 custom = 3;
}

// integer counter pattern
message PatternPrefixConfigAutoFieldTestCounter {

  // Description missing in models
  // default = 0
  optional uint32 start = 1;

  // Description missing in models
  // default = 1
  optional uint32 step = 2;

  // Description missing in models
  // default = 1
  optional uint32 count = 3;
}

// TBD
message PatternPrefixConfigAutoFieldTest {

  message Choice {
    enum Enum {
      unspecified = 0;
      value = 2;
      values = 3;
      auto = 1;
      increment = 4;
      decrement = 5;
    }
  }
  // Description missing in models
  // default = Choice.Enum.auto
  optional Choice.Enum choice = 1;

  // Description missing in models
  // default = 0
  optional uint32 value = 2;

  // Description missing in models
  // default = [0]
  repeated uint32 values = 3;

  // The OTG implementation can provide a system generated
  // value for this property. If the OTG is unable to generate a value
  // the default value must be used.
  // default = 0
  optional uint32 auto = 4;

  // Description missing in models
  optional PatternPrefixConfigAutoFieldTestCounter increment = 6;

  // Description missing in models
  optional PatternPrefixConfigAutoFieldTestCounter decrement = 7;
}

// ipv4 counter pattern
message PatternIpv4PatternIpv4Counter {

  // Description missing in models
  // default = 0.0.0.0
  optional string start = 1;

  // Description missing in models
  // default = 0.0.0.1
  optional string step = 2;

  // Description missing in models
  // default = 1
  optional uint32 count = 3;
}

// TBD
message PatternIpv4PatternIpv4 {

  message Choice {
    enum Enum {
      unspecified = 0;
      value = 2;
      values = 3;
      increment = 4;
      decrement = 5;
    }
  }
  // Description missing in models
  // default = Choice.Enum.value
  optional Choice.Enum choice = 1;

  // Description missing in models
  // default = 0.0.0.0
  optional string value = 2;

  // Description missing in models
  // default = ['0.0.0.0']
  repeated string values = 3;

  // Description missing in models
  optional PatternIpv4PatternIpv4Counter increment = 5;

  // Description missing in models
  optional PatternIpv4PatternIpv4Counter decrement = 6;
}

// ipv6 counter pattern
message PatternIpv6PatternIpv6Counter {

  // Description missing in models
  // default = ::
  optional string start = 1;

  // Description missing in models
  // default = ::1
  optional string step = 2;

  // Description missing in models
  // default = 1
  optional uint32 count = 3;
}

// TBD
message PatternIpv6PatternIpv6 {

  message Choice {
    enum Enum {
      unspecified = 0;
      value = 2;
      values = 3;
      increment = 4;
      decrement = 5;
    }
  }
  // Description missing in models
  // default = Choice.Enum.value
  optional Choice.Enum choice = 1;

  // Description missing in models
  // default = ::
  optional string value = 2;

  // Description missing in models
  // default = ['::']
  repeated string values = 3;

  // Description missing in models
  optional PatternIpv6PatternIpv6Counter increment = 5;

  // Description missing in models
  optional PatternIpv6PatternIpv6Counter decrement = 6;
}

// mac counter pattern
message PatternMacPatternMacCounter {

  // Description missing in models
  // default = 00:00:00:00:00:00
  optional string start = 1;

  // Description missing in models
  // default = 00:00:00:00:00:01
  optional string step = 2;

  // Description missing in models
  // default = 1
  optional uint32 count = 3;
}

// TBD
message PatternMacPatternMac {

  message Choice {
    enum Enum {
      unspecified = 0;
      value = 2;
      values = 3;
      auto = 1;
      increment = 4;
      decrement = 5;
    }
  }
  // Description missing in models
  // default = Choice.Enum.auto
  optional Choice.Enum choice = 1;

  // Description missing in models
  // default = 00:00:00:00:00:00
  optional string value = 2;

  // Description missing in models
  // default = ['00:00:00:00:00:00']
  repeated string values = 3;

  // The OTG implementation can provide a system generated
  // value for this property. If the OTG is unable to generate a value
  // the default value must be used.
  // default = 00:00:00:00:00:00
  optional string auto = 4;

  // Description missing in models
  optional PatternMacPatternMacCounter increment = 6;

  // Description missing in models
  optional PatternMacPatternMacCounter decrement = 7;
}

// integer counter pattern
message PatternIntegerPatternIntegerCounter {

  // Description missing in models
  // default = 0
  optional uint32 start = 1;

  // Description missing in models
  // default = 1
  optional uint32 step = 2;

  // Description missing in models
  // default = 1
  optional uint32 count = 3;
}

// TBD
message PatternIntegerPatternInteger {

  message Choice {
    enum Enum {
      unspecified = 0;
      value = 2;
      values = 3;
      increment = 4;
      decrement = 5;
    }
  }
  // Description missing in models
  // default = Choice.Enum.value
  optional Choice.Enum choice = 1;

  // Description missing in models
  // default = 0
  optional uint32 value = 2;

  // Description missing in models
  // default = [0]
  repeated uint32 values = 3;

  // Description missing in models
  optional PatternIntegerPatternIntegerCounter increment = 5;

  // Description missing in models
  optional PatternIntegerPatternIntegerCounter decrement = 6;
}

// TBD
message PatternChecksumPatternChecksum {

  message Choice {
    enum Enum {
      unspecified = 0;
      generated = 1;
      custom = 2;
    }
  }
  // The type of checksum
  // default = Choice.Enum.generated
  optional Choice.Enum choice = 1;

  message Generated {
    enum Enum {
      unspecified = 0;
      good = 1;
      bad = 2;
    }
  }
  // A system generated checksum value
  // default = Generated.Enum.good
  optional Generated.Enum generated = 2;

  // A custom checksum value
  optional uint32 custom = 3;
}

// Version details
message Version {

  // Version of API specification
  // default =
  optional string api_spec_version = 1;

  // Version of SDK generated from API specification
  // default =
  optional string sdk_version = 2;

  // Version of application consuming or serving the API
  // default =
  optional string app_version = 3;
}

// Success warning payload similar to otg Success
message Warnings {
  WarningDetails warning_details = 1;
}

message SetConfigRequest {
  PrefixConfig prefix_config = 1;
}

message UpdateConfigurationRequest {
  UpdateConfig update_config = 1;
}
message SetConfigResponse {
  bytes response_bytes = 1;
}

message UpdateConfigurationResponse {
  PrefixConfig prefix_config = 1;
}

message GetConfigResponse {
  PrefixConfig prefix_config = 1;
}


message GetMetricsRequest {
  MetricsRequest metrics_request = 1;
}
message GetMetricsResponse {
  Metrics metrics = 1;
}

message GetWarningsResponse {
  WarningDetails warning_details = 1;
}

message ClearWarningsResponse {
  string string = 1;
}


message PostRootResponseRequest {
  ApiTestInputBody api_test_input_body = 1;
}
message GetRootResponseResponse {
  CommonResponseSuccess common_response_success = 1;
}

message DummyResponseTestResponse {
  string string = 1;
}

message PostRootResponseResponse {
  CommonResponseSuccess common_response_success = 1;
}

message GetAllItemsResponse {
  ServiceAbcItemList service_abc_item_list = 1;
}

message GetSingleItemResponse {
  ServiceAbcItem service_abc_item = 1;
}

message GetSingleItemLevel2Response {
  ServiceAbcItem service_abc_item = 1;
}

message GetVersionResponse {
  Version version = 1;
}


// Description missing in models
//
// For all RPCs defined in this service, API Server SHOULD provide JSON
// representation of `Error` message as an error string upon failure, ensuring
// name of enum constants (instead of value) for `kind` property is present
// in the representation
service Openapi {

  // Sets configuration resources.
  rpc SetConfig(SetConfigRequest) returns (SetConfigResponse);
  // Deprecated: please use post instead
  // 
  // Sets configuration resources.
  rpc UpdateConfiguration(UpdateConfigurationRequest) returns (UpdateConfigurationResponse);
  // Gets the configuration resources.
  rpc GetConfig(google.protobuf.Empty) returns (GetConfigResponse);
  // Gets metrics.
  rpc GetMetrics(GetMetricsRequest) returns (GetMetricsResponse);
  // Gets warnings.
  rpc GetWarnings(google.protobuf.Empty) returns (GetWarningsResponse);
  // Clears warnings.
  rpc ClearWarnings(google.protobuf.Empty) returns (ClearWarningsResponse);
  // simple GET api with single return type
  rpc GetRootResponse(google.protobuf.Empty) returns (GetRootResponseResponse);
  // Description missing in models
  rpc DummyResponseTest(google.protobuf.Empty) returns (DummyResponseTestResponse);
  // simple POST api with single return type
  rpc PostRootResponse(PostRootResponseRequest) returns (PostRootResponseResponse);
  // return list of some items
  rpc GetAllItems(google.protobuf.Empty) returns (GetAllItemsResponse);
  // return single item
  rpc GetSingleItem(google.protobuf.Empty) returns (GetSingleItemResponse);
  // return single item
  rpc GetSingleItemLevel2(google.protobuf.Empty) returns (GetSingleItemLevel2Response);
  // Description missing in models
  rpc GetVersion(google.protobuf.Empty) returns (GetVersionResponse);
}
